src/hotspot/share/gc/shared/stringdedup/stringDedupTable.cpp

```diff
@@ -25,631 +25,760 @@
-    Atomic::add(&_entries_removed, removed);
-  }
-}
-
-uintx StringDedupTable::unlink_or_oops_do(StringDedupUnlinkOrOopsDoClosure* cl,
-                                          size_t partition_begin,
-                                          size_t partition_end,
-                                          uint worker_id) {
-  uintx removed = 0;
-  for (size_t bucket = partition_begin; bucket < partition_end; bucket++) {
-    StringDedupEntry** entry = _table->bucket(bucket);
-    while (*entry != NULL) {
-      oop* p = (oop*)(*entry)->obj_addr();
-      if (cl->is_alive(*p)) {
-        cl->keep_alive(p);
-        if (is_resizing()) {
-          // We are resizing the table, transfer entry to the new table
-          _table->transfer(entry, _resized_table);
-        } else {
-          if (is_rehashing()) {
-            // We are rehashing the table, rehash the entry but keep it
-            // in the table. We can't transfer entries into the new table
-            // at this point since we don't have exclusive access to all
-            // destination partitions. finish_rehash() will do a single
-            // threaded transfer of all entries.
-            typeArrayOop value = (typeArrayOop)*p;
-            bool latin1 = (*entry)->latin1();
-            unsigned int hash = hash_code(value, latin1);
-            (*entry)->set_hash(hash);
-          }
-
-          // Move to next entry
-          entry = (*entry)->next_addr();
-        }
-      } else {
-        // Not alive, remove entry from table
-        _table->remove(entry, worker_id);
-        removed++;
-      }
+// Try to look up the string's value array in the shared string table.  This
+// is only worthwhile if sharing is enabled, both at build-time and at
+// runtime.  But it's complicated because we can't trust the is_latin1 value
+// of the string we're deduplicating.  GC requests can provide us with
+// access to a String that is incompletely constructed; the value could be
+// set before the coder.
+bool StringDedup::Table::try_deduplicate_shared(oop java_string) {
+  typeArrayOop value = java_lang_String::value(java_string);
+  assert(value != nullptr, "precondition");
+  assert(TypeArrayKlass::cast(value->klass())->element_type() == T_BYTE, "precondition");
+  int length = value->length();
+  static_assert(sizeof(jchar) == 2 * sizeof(jbyte), "invariant");
+  assert(((length & 1) == 0) || CompactStrings, "invariant");
+  if ((length & 1) == 0) {
+    // If the length of the byte array is even, then the value array could be
+    // either non-latin1 or a compact latin1 that happens to have an even length.
+    // For the former case we want to look for a matching shared string.  But
+    // for the latter we can still do a lookup, treating the value array as
+    // non-latin1, and deduplicating if we find a match.  For deduplication we
+    // only care if the arrays consist of the same sequence of bytes.
+    const jchar* chars = static_cast<jchar*>(value->base(T_CHAR));
```

Comment:
The encoding of the shared strings always match CompactStrings. Otherwise the CDS archive will fail to map. E.g.,

```
$ java -XX:-CompactStrings -Xshare:dump 
$ java -XX:+CompactStrings -Xlog:cds -version
...
[0.032s][info][cds] UseSharedSpaces: The shared archive file's CompactStrings 
                    setting (disabled) does not equal the current CompactStrings setting (enabled).
[0.032s][info][cds] UseSharedSpaces: Unable to map shared spaces
...
```

So you can avoid this `if` block when `CompactStrings == true`. The `!java_lang_String::is_latin1(found)` check below can be changed to an assert.

Also, I think we need an explicit test case where you'd return `true` at line 564. I can write a new test case and email to you. I think it will involve dumping an archive with `-XX:-CompactStrings`.

Reply:
Oh, I realized that my suggestion above is wrong. When `CompactStrings==true`, you can still have a string whose coder is UTF16:

https://github.com/openjdk/jdk/blob/75a2354dc276e107d64516d20fc72bc7ef3d5f86/src/java.base/share/classes/java/lang/String.java#L343-L351

Reply:
Correct.